Fall back to direct cache write if tempfile creation on the same fs fails

myzhang1029 opened the pull request:
Please see #2288 for the discussion.

myzhang1029 committed:
Skip cache output to /dev/null

Signed-off-by: Zhang Maiyun <me@maiyun.me>

myzhang1029 committed:
Fall back to direct write if tmp creation fails

Fixes #2288

The new logic tries to write to a temp file and atomically move it as
before, but if a temp file cannot be created, it falls back to writing
directly to the final location. In the latter case, we also ignore
errors from `set_file_mode`.

Signed-off-by: Zhang Maiyun <me@maiyun.me>

AJIOB commented on src/cache/cache.rs:
```diff
@@ -207,22 +208,41 @@ impl CacheRead {
+                        // Fall back to writing directly to the final location
+                        warn!("Failed to create temp file on the same file system: {e}");
+                        let mut f = File::create(&path)?;
+                        // `optional`` is false in this branch, so do not ignore errors
```
```suggestion
                        // `optional` is false in this branch, so do not ignore errors
```
nit: looks like a typo

myzhang1029 committed:
Fix typo

Co-authored-by: Alex Overchenko <aleksandr9809@gmail.com>

sylvestre commented:
would it be possible to add a test to make sure we don't regress in the future?
thanks

myzhang1029 committed:
Test for cache extraction to /dev/(null|fd)

myzhang1029 commented:
Rebased on `main` and added two tests.

sylvestre commented:
could you please add high level tests in https://github.com/mozilla/sccache/blob/main/tests/system.rs too ? thanks

like
https://github.com/mozilla/sccache/blob/7025295445c607dac66843c572a2a8adc25e8136/tests/system.rs#L1825

myzhang1029 commented:
@sylvestre Sorry for the delayed response. I am planning to test this by adding a function like `test_basic_compile` in `system.rs` which
1. Compiles a file into `/dev/null`
2. Repeats this
3. Check that both commands succeed and that #miss and #hit are both 1.
4. Repeat this for `RawFd`, just like 6014307
Does this sound good to you?

myzhang1029 committed:
High-level tests into /dev/(null|stdout)

myzhang1029 commented:
On a second thought, I have two questions:

I think I named the PR a bit too narrow. Maybe like 
> Fall back to direct cache write if parent directory is not writable

I was think of reframing the PR this way, because this way, we don't have to hardcode `/dev` anymore. We can instead:
1. run the first compilation command
2. `chmod u-w` the destination dir
3. run the second (cached) pass

myzhang1029 commented on src/cache/cache.rs:
```diff
@@ -207,22 +208,41 @@ impl CacheRead {
                 optional,
             } in objects
             {
+                if path == Path::new("/dev/null") {
```
Sorry I do not have a MSVC environment to test this so I wasn't sure. 

For `gcc`, it seems to be complicated by `gcc` appending `.exe` automatically:
- `-o NUL`: it gives me `NUL.exe`
- `-o NUL.`: `ld.exe: cannot open output file NUL.: Invalid argument`
- `-Wl,-o,NUL`: `ld.exe: final link failed: file truncated`
- `-c -o NUL`: nothing (good)
- `-c -o NUL.o`: `NUL.o` (as expected)

AJIOB replied:
Looks like we should support `-c -o NUL` as a base one, isn't it?

P.S. I don't think we really support `-o NUL` variant with the compiling & linking together: @Xuanwo , do we?

myzhang1029 commented:
Sorry for the delay as my area was hit by some severe weather.

This time is largely just rebasing the code and resolving conflicts, but this version should already fix the `openssl` building error you mentioned. What is left is to figure out the tests properly.

myzhang1029 committed:
Skip output to `NUL` on Windows targets

myzhang1029 committed:
Portable unit test: create ro dir and write there

In addition, when `/dev/fd/{fd}` doesn't actually exist (e.g. FreeBSD
without `fdescfs`), skip the test to extract to `/dev/fd/{fd}`.

Signed-off-by: Zhang Maiyun <me@maiyun.me>

AJIOB reviewed (approved):
All looks fine for me

AJIOB commented:
Hi @drahnr ,

Is it OK for you?

myzhang1029 commented:
another rebase onto main, mainly for checking compatibility with #2581

drahnr commented on src/cache/cache_io.rs:
```diff
@@ -143,22 +143,52 @@ impl CacheRead {
+                if path == Path::new("NUL") {
+                    debug!("Skipping output to NUL");
+                    continue;
+                }
```
Could we move the check into a function so the target specifics don't appear in scope?

myzhang1029 replied:
Done

drahnr commented on tests/system.rs:
```diff
@@ -265,6 +267,110 @@ fn test_basic_compile(compiler: Compiler, tempdir: &Path) {
+
+#[cfg(not(unix))]
+fn test_basic_compile_into_dev_stdout(_: Compiler, _: &Path) {
+    warn!("Not unix, skipping tests with /dev/stdout");
```
Please downgrade to an info level log.

myzhang1029 committed:
Change warn to info in test

myzhang1029 committed:
Extract null path from logic

myzhang1029 committed:
Gate windows/unix null paths in integration tests

myzhang1029 committed:
Test and fix NUL handling on windows

myzhang1029 commented:
I got my hands on a windows machine and it turned out I simplified the NUL mess. [Now fixed](https://github.com/mozilla/sccache/pull/2369/changes#diff-cfa6a2b9469ebacdfb8512b0b7486cd85707ad354d82887c8b70fe20e136c859R210).

myzhang1029 commented:
@drahnr The testsuites seem to work on my FreeBSD VM. Could you please have a look? Thanks!

myzhang1029 commented:
@drahnr @sylvestre Hi Both,

How do you feel about merging this patch?
